In [2]:
from snowshovel.io import read_single_alert, parse_alerts
import pandas as pd
from tqdm import tqdm
from pathlib import Path

%matplotlib inline

In [23]:
df = pd.read_table("C:\\Users\\oakpa\\Desktop\\Al_Summer\\milliquas.txt")
df.head()

0.0006286  35.5178439  SDSS J000000.15+353104.2  Q    17.93 19.00 pG  - -  0.845 DR16Q  DR16                                                                                                      
0    0.0006575 -20.0743440  UVQSJ000000.15-200427...                                                                                                                                                  
1    0.0014503  31.0570374  SDSS J000000.33+31032...                                                                                                                                                  
2    0.0015265  27.7322834  SDSS J000000.36+27435...                                                                                                                                                  
3    0.0015354   7.0641294  SDSS J000000.36+07035...                                                                                                                                                  
4    0.0018983  17.7737392  SDSS J000000.45+17462...

Set up the config here for running the notebook

In [6]:
# Set this to the directory containing the unzipped avro alerts
data_dir = Path("/Users/oakpa/Desktop/Al_Summer/Avro_Files")

# Set the number of alerts to read
n_to_read = 1000

## What is an avro alert?

We can firstly look at a single alert, to see what it looks like:

In [7]:
alert_path = data_dir.joinpath("2350249170015010010.avro")
print(f"We will look at the avro alert with path '{alert_path}'")

We will look at the avro alert with path '\Users\oakpa\Desktop\Al_Summer\Avro_Files\2350249170015010010.avro'


In [8]:
alert = read_single_alert(alert_path)
print("An avro alert is just a python dictionary, as you can see below:")
alert

An avro alert is just a python dictionary, as you can see below:


{'jd': 2460104.7491782,
 'fid': 1,
 'pid': 2350249170015,
 'diffmaglim': 20.196775436401367,
 'pdiffimfilename': 'ztf_20230609249155_000326_zg_c01_o_q1_scimrefdiffimg.fits',
 'programpi': 'Kulkarni',
 'programid': 1,
 'candid': 2350249170015010010,
 'isdiffpos': 'f',
 'tblid': 10,
 'nid': 2350,
 'rcid': 0,
 'field': 326,
 'xpos': 1293.8194580078125,
 'ypos': 841.447509765625,
 'ra': 230.9148832,
 'dec': -19.2385929,
 'magpsf': 17.376073837280273,
 'sigmapsf': 0.047354377806186676,
 'chipsf': 14.490418434143066,
 'magap': 17.40329933166504,
 'sigmagap': 0.03269999846816063,
 'distnr': 0.16106265783309937,
 'magnr': 15.559000015258789,
 'sigmagnr': 0.013000000268220901,
 'chinr': 0.6050000190734863,
 'sharpnr': -0.017999999225139618,
 'sky': 0.5778443217277527,
 'magdiff': 0.027225999161601067,
 'fwhm': 2.9700000286102295,
 'classtar': 0.9950000047683716,
 'mindtoedge': 841.447509765625,
 'magfromlim': 2.793475389480591,
 'seeratio': 0.9369040727615356,
 'aimage': 0.8059999942779541,
 'b

## Handling many avro alerts:

That was just one avro alert, but our data will contain many sources, each with their own alert. python dictionaries are not very convenient for a large number of entries. Instead, we can use a function to read in a number of the alerts and convert them to a pandas dataframe.

In [9]:
alerts_df = parse_alerts(data_dir=data_dir, n_to_read=n_to_read)
print("Here you can see a pandas dataframe. The avro alert above is the first row in the table.")
alerts_df

100%|██████████████████████████████████████████████████████████████████████████████| 1000/1000 [01:21<00:00, 12.24it/s]


Here you can see a pandas dataframe. The avro alert above is the first row in the table.


jd  fid            pid  diffmaglim  \
0    2.460105e+06    1  2350249170015   20.196775   
1    2.460105e+06    1  2350249170015   20.196775   
2    2.460105e+06    1  2350249170015   20.196775   
3    2.460105e+06    1  2350249170015   20.196775   
4    2.460105e+06    1  2350249170015   20.196775   
..            ...  ...            ...         ...   
995  2.460105e+06    1  2350249650415   20.518072   
996  2.460105e+06    1  2350249650415   20.518072   
997  2.460105e+06    1  2350249650515   20.420591   
998  2.460105e+06    1  2350249650515   20.420591   
999  2.460105e+06    1  2350249650515   20.420591   

                                       pdiffimfilename programpi  programid  \
0    ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
1    ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
2    ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
3    ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
4    ztf_20230609249155_000326_zg_c01_o_q1_scimrefd...  Kulkarni          1   
..                                                 ...       ...        ...   
995  ztf_20230609249641_000427_zg_c02_o_q1_scimrefd...  Kulkarni          1   
996  ztf_20230609249641_000427_zg_c02_o_q1_scimrefd...  Kulkarni          1   
997  ztf_20230609249641_000427_zg_c02_o_q2_scimrefd...  Kulkarni          1   
998  ztf_20230609249641_000427_zg_c02_o_q2_scimrefd...  Kulkarni          1   
999  ztf_20230609249641_000427_zg_c02_o_q2_scimrefd...  Kulkarni          1   

                  candid isdiffpos  tblid  ...      zpmed  clrmed    clrrms  \
0    2350249170015010010         f     10  ...  26.086000   0.554  0.230397   
1    2350249170015010011         f     11  ...  26.086000   0.554  0.230397   
2    2350249170015010013         f     13  ...  26.086000   0.554  0.230397   
3    2350249170015010023         f     23  ...  26.086000   0.554  0.230397   
4    2350249170015010024         f     24  ...  26.086000   0.554  0.230397   
..                   ...       ...    ...  ...        ...     ...       ...   
995  2350249650415015015         t     15  ...  26.273001   0.569  0.284820   
996  2350249650415015017         t     17  ...  26.273001   0.569  0.284820   
997  2350249650515010002         f      2  ...  26.290001   0.560  0.299331   
998  2350249650515010003         f      3  ...  26.290001   0.560  0.299331   
999  2350249650515010007         f      7  ...  26.290001   0.560  0.299331   

      neargaia  neargaiabright    maggaia  maggaiabright  exptime       drb  \
0     0.228274     -999.000000  15.315500    -999.000000     30.0  0.999971   
1     0.066207     -999.000000  14.556261    -999.000000     30.0  0.999255   
2     0.175491     -999.000000  16.767410    -999.000000     30.0  1.000000   
3     0.398773     -999.000000  16.487196    -999.000000     30.0  1.000000   
4     0.105536       52.636383  18.771164      13.606915     30.0  0.999883   
..         ...             ...        ...            ...      ...       ...   
995  38.405056     -999.000000  19.375727    -999.000000     30.0  0.999941   
996   0.110178       74.885857  19.273678      13.588749     30.0  0.999999   
997   0.184441       84.466194  15.270944      13.869880     30.0  0.999985   
998   0.112007     -999.000000  15.355320    -999.000000     30.0  1.000000   
999   0.344461       62.260994  15.493838      12.657384     30.0  1.000000   

     drbversion  
0         d6_m7  
1         d6_m7  
2         d6_m7  
3         d6_m7  
4         d6_m7  
..          ...  
995       d6_m7  
996       d6_m7  
997       d6_m7  
998       d6_m7  
999       d6_m7  

[1000 rows x 103 columns]

The table above contains many rows and columns. It can be hard to see what the important information. Let's just display a few relevant columns, rather than all 103.

In [10]:
useful_columns = [
    "isdiffpos", # Is the detection postive (more flux in science image than reference), or negative (less flux)?
    "ra", # What is the sky position in Right Ascension?
    "dec", # What is the sky position in Declination?
    "magpsf", # What is the magnitude (brightness) of the detection?
    "sigmapsf", # What is the uncertainty on the magnitude?
    "distnr", # How close is the detection to a known source in the reference image?
    "ndethist", # How many times has this object been detected?
]
slim_alerts_df = alerts_df[useful_columns]
slim_alerts_df

isdiffpos          ra        dec     magpsf  sigmapsf     distnr  ndethist
0           f  230.914883 -19.238593  17.376074  0.047354   0.161063       620
1           f  230.630090 -19.273569  17.239075  0.078380   0.173305       639
2           f  231.014147 -19.282204  18.312075  0.061234   0.134472       611
3           f  230.909926 -19.485533  18.983074  0.085728   0.424015       721
4           f  230.435069 -19.499763  20.108074  0.164924   0.145266        65
..        ...         ...        ...        ...       ...        ...       ...
995         t  222.018785  -5.363314  20.120691  0.146097  12.357216         1
996         t  222.300244  -5.390949  19.616692  0.098350   0.192763       134
997         f  220.804638  -4.758334  19.417479  0.189056   0.228310       104
998         f  220.716001  -4.964255  18.074478  0.091027   0.163497       513
999         f  221.481146  -5.403756  19.339479  0.121661   0.341977       349

[1000 rows x 7 columns]

In [11]:
# Bad subtractions are often not repeated
# Let's restrict ourselves to sources which are detected at least twice
mask = slim_alerts_df["ndethist"] > 1
slim_alerts_df = slim_alerts_df[mask].reset_index()
print(f"There are {len(slim_alerts_df)} alerts passing our cut")
slim_alerts_df

There are 409 alerts passing our cut


index isdiffpos          ra        dec     magpsf  sigmapsf    distnr  \
0        0         f  230.914883 -19.238593  17.376074  0.047354  0.161063   
1        1         f  230.630090 -19.273569  17.239075  0.078380  0.173305   
2        2         f  231.014147 -19.282204  18.312075  0.061234  0.134472   
3        3         f  230.909926 -19.485533  18.983074  0.085728  0.424015   
4        4         f  230.435069 -19.499763  20.108074  0.164924  0.145266   
..     ...       ...         ...        ...        ...       ...       ...   
404    993         t  222.257711  -5.459212  19.301691  0.095494  9.619986   
405    996         t  222.300244  -5.390949  19.616692  0.098350  0.192763   
406    997         f  220.804638  -4.758334  19.417479  0.189056  0.228310   
407    998         f  220.716001  -4.964255  18.074478  0.091027  0.163497   
408    999         f  221.481146  -5.403756  19.339479  0.121661  0.341977   

     ndethist  
0         620  
1         639  
2         611  
3         721  
4          65  
..        ...  
404         2  
405       134  
406       104  
407       513  
408       349  

[409 rows x 8 columns]